src/Adapter/Category/CommandHandler/AbstractDeleteCategoryHandler.php

```diff
@@ -0,0 +1,75 @@
+<?php
+/**
+ * 2007-2018 PrestaShop.
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\Category\CommandHandler;
+
+use PrestaShop\PrestaShop\Core\Domain\Category\ValueObject\CategoryDeleteMode;
+use Product;
+use Shop;
+
+/**
+ * Class AbstractDeleteCategoryHandler.
+ */
+abstract class AbstractDeleteCategoryHandler
+{
+    /**
+     * Handle products category after its deletion.
+     *
+     * @param int $parentCategoryId
+     * @param CategoryDeleteMode $mode
+     */
+    protected function handleProductsUpdate($parentCategoryId, CategoryDeleteMode $mode)
+    {
+        $productsWithoutCategory = \Db::getInstance()->executeS('
```

Comment:
This might not be the right PR to introduce it
but should'nt we create a LegacyQueryBuilder to handle these kind of queries ?

```
$legacyQueryBuilder = new LegacyQueryBuilder($this->getContext(), $this->getWhateverNeeded() ...);

$subQuery = new LegacyQueryBuilder(...);
$subQuery->select(1)->from('category_product cp')->where('cp.id_product = p.id_product');

$legacyQueryBuilder->select('p.id_product')
                   ->from('product p')
                   ->addShopAssociation('p')
                   ->where('NOT EXISTS '.$subQuery->getSql());

$result = \Db::getInstance()->executeS($legacyQueryBuilder->getSql());
```

Reply:
@mickaelandrieu @eternoendless

Reply:
but why? would that bring any value? seems that using plain sql provides less coupling with legacy code. :thinking: